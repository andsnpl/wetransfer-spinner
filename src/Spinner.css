.spinner {
  animation: rotate 3s linear infinite;
}

.spinner-reverse {
  animation: rotate 3s linear infinite reverse;
}

.spinner-paused {
  animation-play-state: paused;
}

@keyframes rotate {
  from { transform: rotate(0); }
  to { transform: rotate(360deg); }
}