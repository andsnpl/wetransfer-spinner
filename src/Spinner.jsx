import * as React from "react";
import "./Spinner.css";
import CircularGauge from "./CircularGauge";

function Spinner({ state, value }) {
  value = Math.max(Math.min(value, 100), 0);
  const gaugeValue = Math.max(value, 1);
  const [outerClasses, innerClasses] = {
    READY: ["", ""],
    ACTIVE: ["spinner", "spinner-reverse"],
    PAUSED: ["spinner spinner-paused", "spinner-reverse spinner-paused"]
  }[state];
  return (
    <>
      <div className={outerClasses}>
        <CircularGauge value={gaugeValue} origin="90deg">
          {/**
           * Replacing the default readout for two reasons:
           * 1. We should display 0% if value is 0, even though we force a
           *    minimum 1% arc on the gauge.
           * 2. Gauge shouldn't know anything about the spinner's rotation, so
           *    we rotate the outer div and then reverse it for the readout.
           */}
          <div className={innerClasses}>
            <CircularGauge.Readout value={value} />
          </div>
        </CircularGauge>
      </div>
    </>
  );
}

export default Spinner;
