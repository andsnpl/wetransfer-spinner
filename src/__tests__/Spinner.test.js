import { render } from "@testing-library/react";
import Spinner from "../Spinner";

describe("Spinner", () => {
  it("should render in the READY state", () => {
    const { container } = render(<Spinner state="READY" value={0} />);
    expect(container).toMatchSnapshot();
  });

  it("should render in the ACTIVE state", () => {
    const { container } = render(<Spinner state="ACTIVE" value={0} />);
    expect(container).toMatchSnapshot();
  });

  it("should render in the PAUSED state", () => {
    const { container } = render(<Spinner state="PAUSED" value={0} />);
    expect(container).toMatchSnapshot();
  });

  it("should render when partially completed", () => {
    const { container } = render(<Spinner state="ACTIVE" value={66} />);
    expect(container).toMatchSnapshot();
  });
});
