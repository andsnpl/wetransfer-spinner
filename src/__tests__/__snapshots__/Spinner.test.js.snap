// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner should render in the ACTIVE state 1`] = `
<div>
  <div
    class="spinner"
  >
    <div
      class="gauge"
    >
      <div
        aria-hidden="true"
        class="gauge-circle"
        style="background-color: rgb(64, 159, 255); transform: rotate(90deg);"
      >
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(3.6deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(90deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(180deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(270deg);"
        />
        <div
          class="gauge-endcap-container"
          style="transform: rotate(0deg);"
        >
          <div
            class="gauge-endcap"
            style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
          />
        </div>
        <div
          class="gauge-endcap-container"
          style="transform: rotate(3.6deg);"
        >
          <div
            class="gauge-endcap"
            style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
          />
        </div>
      </div>
      <div
        class="gauge-center"
        style="background-color: rgb(255, 255, 255);"
      >
        <div
          class="spinner-reverse"
        >
          <div
            class="gauge-readout"
          >
            <span
              class="gauge-value"
            >
              0
            </span>
            <span
              class="gauge-percent-sign"
            >
              %
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Spinner should render in the PAUSED state 1`] = `
<div>
  <div
    class="spinner spinner-paused"
  >
    <div
      class="gauge"
    >
      <div
        aria-hidden="true"
        class="gauge-circle"
        style="background-color: rgb(64, 159, 255); transform: rotate(90deg);"
      >
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(3.6deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(90deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(180deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(270deg);"
        />
        <div
          class="gauge-endcap-container"
          style="transform: rotate(0deg);"
        >
          <div
            class="gauge-endcap"
            style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
          />
        </div>
        <div
          class="gauge-endcap-container"
          style="transform: rotate(3.6deg);"
        >
          <div
            class="gauge-endcap"
            style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
          />
        </div>
      </div>
      <div
        class="gauge-center"
        style="background-color: rgb(255, 255, 255);"
      >
        <div
          class="spinner-reverse spinner-paused"
        >
          <div
            class="gauge-readout"
          >
            <span
              class="gauge-value"
            >
              0
            </span>
            <span
              class="gauge-percent-sign"
            >
              %
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Spinner should render in the READY state 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="gauge"
    >
      <div
        aria-hidden="true"
        class="gauge-circle"
        style="background-color: rgb(64, 159, 255); transform: rotate(90deg);"
      >
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(3.6deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(90deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(180deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(270deg);"
        />
        <div
          class="gauge-endcap-container"
          style="transform: rotate(0deg);"
        >
          <div
            class="gauge-endcap"
            style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
          />
        </div>
        <div
          class="gauge-endcap-container"
          style="transform: rotate(3.6deg);"
        >
          <div
            class="gauge-endcap"
            style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
          />
        </div>
      </div>
      <div
        class="gauge-center"
        style="background-color: rgb(255, 255, 255);"
      >
        <div
          class=""
        >
          <div
            class="gauge-readout"
          >
            <span
              class="gauge-value"
            >
              0
            </span>
            <span
              class="gauge-percent-sign"
            >
              %
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Spinner should render when partially completed 1`] = `
<div>
  <div
    class="spinner"
  >
    <div
      class="gauge"
    >
      <div
        aria-hidden="true"
        class="gauge-circle"
        style="background-color: rgb(64, 159, 255); transform: rotate(90deg);"
      >
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(237.6deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(237.6deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(237.6deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(226, 228, 230); transform: rotate(270deg);"
        />
        <div
          class="gauge-background-panel"
          style="background-color: rgb(64, 159, 255); transform: rotate(0deg);"
        />
        <div
          class="gauge-endcap-container"
          style="transform: rotate(0deg);"
        >
          <div
            class="gauge-endcap"
            style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
          />
        </div>
        <div
          class="gauge-endcap-container"
          style="transform: rotate(237.6deg);"
        >
          <div
            class="gauge-endcap"
            style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
          />
        </div>
      </div>
      <div
        class="gauge-center"
        style="background-color: rgb(255, 255, 255);"
      >
        <div
          class="spinner-reverse"
        >
          <div
            class="gauge-readout"
          >
            <span
              class="gauge-value"
            >
              66
            </span>
            <span
              class="gauge-percent-sign"
            >
              %
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
